Fixed issue #496 - sanitize/truncate bug

LordSputnik opened the pull request:
Added loop to iterate over sanitize/truncate until stable. Enabled test_truncation_does_not_conflict_with_replacement test. See discussion in #496.

sampsyo commented on beets/library.py:
```diff
@@ -715,6 +715,71 @@ def move(self, copy=False, link=False, basedir=None, with_album=True):
+            while path not in path_candidates:
+                path_candidates.append(path)
+                # Convert back to Unicode with extension removed
+                print(util.displayable_path(path))
```
Stray debugging print.

LordSputnik replied:
Ah yes - will deal with that when I update the PR.

sampsyo commented on beets/library.py:
```diff
@@ -715,6 +715,71 @@ def move(self, copy=False, link=False, basedir=None, with_album=True):
+        while path != path_candidates[-1]:
+            # This will keep sanitizing the path until it's stable, or an
+            # infinite loop appears
+            while path not in path_candidates:
```
Nifty: you're explicitly detecting an infinite loop by checking whether we're backtracking to a previously-seen state.

An even simpler heuristic would probably work for our purposes: e.g., if after replacing/truncating/replacing in a single cycle the path is too long again. It's less exciting than detecting backtracking, but perhaps this would avoid weird situations where this would cause many iterations.

LordSputnik replied:
I expect the loop would converge fairly quickly anyway, so exiting after one cycle would only cause the user's replacements to be honoured less, with little performance increase. In the rare case that the loop cycles more than once, the time taken to perform the iterations is likely to be tiny compared to the overall time of whatever operation they're performing with beets.

sampsyo replied:
Yeah, I totally agree that we should not prematurely optimize. This is quite unlikely to be a bottleneck. :smiley: I was just proposing the simpler one-shot approach for predictability, not really performance. Since we can't prove that weird, unbounded, acyclic fluctuations won't happen, I thought that a simple limit would make it certain that this doesn't run for an arbitrary (finite) amount of time.

I totally get it if you think this version is simpler or more maintainable, though, so I'll defer to your preference.

sampsyo commented on beets/library.py:
```diff
@@ -715,6 +715,71 @@ def move(self, copy=False, link=False, basedir=None, with_album=True):
 
     # Templating.
 
+    def _legalize_partial(self, path, fragment, replacements):
```
Since this is getting rather complex, I propose to break off this method into a pure function so we can test it independently. For this, that would mean passing in `self.path` and the max filename length as arguments; these can then be provided by the `destination` wrapper, which will remain a method.

It might also be useful to give a similar treatment to `_legalize_path` itself. Testability will be really helpful here.

LordSputnik replied:
That sounds good. I did initially think about putting them in utils, alongside sanitize and truncate.

LordSputnik commented:
I've added some replies to your comments - will tweak this PR hopefully at the weekend, or by Tuesday if not then.

sampsyo commented:
Great! Thanks again! If you just leave a comment when you're happy with the next round, I'll get an email to come back and merge.

LordSputnik commented:
Completely forgot I had this open still - will make the necessary changes and close tonight if I get time.

LordSputnik commented:
So, I finally finished this and I'm happy with it now. No stray prints, no loops to get stuck in and predictable behaviour. There are also two tests for it, which both pass.

The new simpler algorithm is to do one pass with sanitize using the user replacements, followed by truncate. Then, a second pass with the user replacements, and a test to see whether further truncation occurred. If it did, remove the user replacements, and sanitize and truncate (based on the assumption that none of the built in replacements will ever increase the path length - it might be good to throw in a test for that, but I can't see a good way to do it).

sampsyo commented on beets/util/__init__.py:
```diff
@@ -545,6 +545,59 @@ def truncate_path(path, length=MAX_FILENAME_LENGTH):
+        # Outputting Unicode.
+        extension = extension.decode('utf8', 'ignore')
+
+    first_stage_path =\
```
Should this be `first_stage_path, _ = ...` to ignore the second returned value?

sampsyo replied:
Oops! Just noticed the `[0]` below. Using the unpacking syntax can be marginally clearer, though.

sampsyo commented on beets/util/__init__.py:
```diff
@@ -545,6 +545,59 @@ def truncate_path(path, length=MAX_FILENAME_LENGTH):
+    pre_truncate_path = path
+    path = truncate_path(path, length)
+
+    return (path, path != pre_truncate_path)
```
Tiny style nit: the parentheses are not needed here.

sampsyo commented on beets/util/__init__.py:
```diff
@@ -545,6 +545,59 @@ def truncate_path(path, length=MAX_FILENAME_LENGTH):
 
+def _legalize_stage(path, replacements, length, extension, fragment):
+    """ Perform a signle stage of legalization of the path (ie. a single
+    sanitization and truncation). Returns the path (unicode if fragment is set,
```
Perhaps, to be slightly more specific:

> (i.e., it sanitizes the Unicode path, encodes it to bytes, adds the extension, and then truncates it)

sampsyo commented on beets/util/__init__.py:
```diff
@@ -545,6 +545,59 @@ def truncate_path(path, length=MAX_FILENAME_LENGTH):
+        first_stage_path, replacements, length, extension, fragment
+    )
+
+    # If the path was once again truncated, discard user replacements.
```
Suggested expansion:

> This is to avoid a "vicious cycle," in which truncation requires sanitation, then that sanitation requires truncation, ad infinitum.

sampsyo commented:
Very nice! I like this version a lot.

So it's still possible for users to write "evil" replacements that increase the length of the path. This version essentially says, "if you do that, then your replacements may not always be obeyed". The only other alternatives I can see would be:

* "If you do that, then you may get filenames that are too long." (the current behavior in beets)
* "If you do that, then beets will print an error message and exit."

In any case, we should probably add some sort of warning to the documentation, right? The policy in this version *can* create invalid paths if the replacement enforces some OS requirement—for example, with trailing whitespace on Windows—and beets will crash later when it tries to create a file with that name. There's no way around something going wrong, though, so the best we can do is provide good documentation.

LordSputnik commented:
Ok, I've updated the function docstrings and switched over to unpacking syntax as suggested.

I think it'd be a good idea to warn users in the documentation, and also to warn the user at runtime if replacements have been ignored. I guess this should be mentioned at https://beets.readthedocs.org/en/v1.3.13/reference/config.html#replace for the documentation, but not sure how the warning should be generated in beets? Do you have some sort of logging system, or would it just be implemented with print?

sampsyo commented:
That sounds like the perfect place for the docs warning.

And yes, there is a logging system. For example: https://github.com/sampsyo/beets/blob/master/beets/library.py#L612
The only trick is that we try to keep the util module log-free (just pure utilities). So we may need to either move some code or return a flag so we can log the message from library.py.

LordSputnik commented:
All done now, think it's ready to merge!

sampsyo commented:
Awesome! This is looking great. I'll merge this.

The next step in this direction will be to address #1533/#1418. The `_legalize_stage` machinery, which is already responsible for both encoding and truncation, should do so now in an encoding-aware way. Hopefully, we can also find a way to avoid the encode-then-decode-again cycle that we have to resort to at the moment.

sampsyo commented:
To be clear: it would be great if, as this continues to evolve, we can make the common case (no truncation) not require encoding, decoding to Unicode again, then re-encoding.

sampsyo commented:
Just to reiterate: :sparkles: THANK YOU :sparkles: for working on such a nasty, deceptively complicated issue. Woohoo!

I'm adding you as a collaborator in case you want to do further maintenance.

LordSputnik commented:
No problem :) I'm especially glad that this is fixed because it was affecting a couple of songs in my library.

Will help out where I can if I have some spare time. I'm planning to write a python module to handle platform-dependency in path names as a result of working on this issue (there doesn't seem to be one!), and that would hopefully solve the unicode character issues, as well as improve the existing solution applied here.

sampsyo commented:
A library like that sounds incredibly useful. Perhaps it would be worthwhile to dovetail with [pathlib](https://pypi.python.org/pypi/pathlib/), which has a (very basic) notion of platform-specific paths?
